src/rgw/rgw_common.cc

```diff
@@ -1467,3 +1478,68 @@ int rgw_parse_op_type_list(const string& str, uint32_t *perm)
   return parse_list_of_flags(op_type_mapping, str, perm);
 }
 
+static int match_internal(const char* pattern, const char* input)
+{
+  if (*pattern == '\0' && *input == '\0')
+      return 1;
+  if (*pattern == '\0' || *input == '\0')
+      return 0;
+  if (*pattern == '*' && *(pattern + 1) == '\0' && *input != '\0')
+      return 1;
+  if (*pattern == '*' && *(pattern + 1) == '\0' && *input == '\0')
+      return 0;
+  if (*pattern == *input || *pattern == '?')
+      return match_internal(pattern + 1, input + 1);
+  if (*pattern == '*')
+      return match_internal(pattern + 1, input) || match_internal(pattern, input + 1);
```

Comment:
Ironically, since these are tailcalls, GCC might even compile it as non-recursive. We can't depend on that.

I for one would be in favor of rewriting all of RadosGW in Scheme, though :)